File: content/snippets/scala3/interview_lcci_04_06_successor.scala
object Lcci0406Successor:
  final case class Node(value: Int, var left: Node | Null = null, var right: Node | Null = null)

  def inorderSuccessor(root0: Node | Null, p: Int): Node | Null =
    var root = root0
    var ans: Node | Null = null
    while (root != null) {
      val cur = root.asInstanceOf[Node]
      if (p < cur.value) {
        ans = cur
        root = cur.left
      } else {
        root = cur.right
      }
    }
    ans

  @main def main(): Unit =
    val root = Node(5, Node(3, Node(2), Node(4)), Node(7))
    val s1 = inorderSuccessor(root, 3)
    val s2 = inorderSuccessor(root, 7)
    println(if (s1 == null) -1 else s1.asInstanceOf[Node].value)
    println(if (s2 == null) -1 else s2.asInstanceOf[Node].value)

File: content/snippets/cpp/interview_lcci_04_06_successor.cpp
#include <iostream>

struct Node { int v; Node* l; Node* r; Node(int x): v(x), l(nullptr), r(nullptr) {} };

Node* inorderSuccessor(Node* root, int p) {
    Node* ans = nullptr;
    while (root) {
        if (p < root->v) { ans = root; root = root->l; }
        else root = root->r;
    }
    return ans;
}

int main() {
    Node* root = new Node(5);
    root->l = new Node(3); root->r = new Node(7);
    root->l->l = new Node(2); root->l->r = new Node(4);
    Node* s1 = inorderSuccessor(root, 3);
    Node* s2 = inorderSuccessor(root, 7);
    std::cout << (s1 ? s1->v : -1) << '\n';
    std::cout << (s2 ? s2->v : -1) << '\n';
}
